Ruby/Ruby.sublime-syntax

```diff
@@ -1345,97 +1345,135 @@ contexts:
-    #     - include: interpolated-ruby
-    #     - include: escaped-char
+    #   push: [heredoc-shell, trailing-heredoc-start]
     - match: (\=)\s*(<<(\w+))
       captures:
         1: keyword.operator.assignment.ruby
         2: punctuation.definition.string.begin.ruby
-      push:
-        - meta_scope: string.unquoted.heredoc.ruby
-        - match: ^\s*\3$
-          scope: punctuation.definition.string.end.ruby
-          pop: true
-        - include: interpolated-ruby
-        - include: escaped-char
+      push: [heredoc-assign, trailing-heredoc-no-embedding-start]
     # heredoc with indented terminator
     - match: '(<<[-~](\w+))'
       scope: punctuation.definition.string.begin.ruby
-      push:
-        - meta_scope: string.unquoted.heredoc.ruby
-        - match: ^\s*\2$
-          scope: punctuation.definition.string.end.ruby
-          pop: true
-        - include: interpolated-ruby
-        - include: escaped-char
+      push: [heredoc-plain, trailing-heredoc-no-embedding-start]
+
+  heredoc-html:
+    - meta_scope: string.unquoted.embedded.html.ruby
+    - meta_content_scope: text.html.embedded.ruby
+    - match: ^\s*\2$
+      scope: punctuation.definition.string.end.ruby
+      pop: true
+    - include: scope:text.html.basic
+    - include: interpolated-ruby
+    - include: escaped-char
+
+  heredoc-sql:
+    - meta_scope: string.unquoted.embedded.sql.ruby
+    - meta_content_scope: text.sql.embedded.ruby
+    - match: ^\s*\2$
+      scope: punctuation.definition.string.end.ruby
+      pop: true
+    - include: scope:source.sql
+    - include: interpolated-ruby
+    - include: escaped-char
+
+  heredoc-css:
+    - meta_scope: string.unquoted.embedded.css.ruby
+    - meta_content_scope: text.css.embedded.ruby
+    - match: ^\s*\2$
+      scope: punctuation.definition.string.end.ruby
+      pop: true
+    - include: 'scope:source.css'
+    - include: interpolated-ruby
+    - include: escaped-char
+
+  heredoc-js:
+    - meta_scope: string.unquoted.embedded.js.ruby
+    - meta_content_scope: text.js.embedded.ruby
+    - match: ^\s*\2$
+      scope: punctuation.definition.string.end.ruby
+      pop: true
+    - include: 'scope:source.js'
+    - include: interpolated-ruby
+    - include: escaped-char
+
+  heredoc-ruby:
+    - meta_scope: string.unquoted.embedded.ruby.ruby
+    - meta_content_scope: text.ruby.embedded.ruby
+    - match: ^\s*\2$
+      scope: punctuation.definition.string.end.ruby
+      pop: true
+    - include: interpolated-ruby
+    - include: escaped-char
+    - include: expressions
+
+  #heredoc-shell:
+  #  - meta_scope: string.unquoted.embedded.shell.ruby
+  #  - meta_content_scope: text.shell.embedded.ruby
+  #  - match: ^\s*\2$
+  #    scope: punctuation.definition.string.end.ruby
+  #    pop: true
+  #  - include: Shell-Unix-Generic.sublime-syntax
+  #  - include: interpolated-ruby
+  #  - include: escaped-char
+
+  # This prevents clear_scopes from applying to the push token
```

Comment:
@wbond what does this mean? Why is this necessary? This context seems useless to me.

I'm trying to implement `clear_scopes` in https://github.com/trishume/syntect and the fact that this context looks useless to me suggests I might be misunderstanding the semantics of either `clear_scopes` or push/pop operations. 

If you could explain to me why this is necessary I would really appreciate it. Also, what is a "push token" and why would `clear_scopes` apply to it?